pkg/admission/ingress/webhook_handler.go

```diff
@@ -0,0 +1,147 @@
+	"bytes"
+	"context"
+	"fmt"
+
+	// corev1 "k8s.io/api/core/v1"
+	// "net/http"
+
+	kuadrantv1 "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/dns"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/traffic"
+	log "github.com/sirupsen/logrus"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	controllertraffic "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/controllers/traffic"
+)
+
+type Handler struct {
+	Decoder    *admission.Decoder
+	Serializer *json.Serializer
+	client.Client
+
+	Hosts        controllertraffic.HostService
+	Certificates controllertraffic.CertificateService
+	HostResolver dns.HostResolver
+}
+
+func CreateHandler(hostService controllertraffic.HostService, certService controllertraffic.CertificateService, hostResolver dns.HostResolver) (admission.Handler, error) {
+	log.Info("Creating handler ")
+	scheme := runtime.NewScheme()
+	if err := networkingv1.AddToScheme(scheme); err != nil {
+		log.Error("Unable add to scheme", err)
+		return nil, err
+	}
+
+	serialiser := json.NewSerializerWithOptions(json.DefaultMetaFactory, scheme, scheme, json.SerializerOptions{})
+
+	log.Info("Creating decoder")
+	decoder, err := admission.NewDecoder(scheme)
+	if err != nil {
+		log.Error("Unable to create decoder", err)
+		return nil, err
+	}
+	return &Handler{
+		Decoder:    decoder,
+		Serializer: serialiser,
+
+		Hosts:        hostService,
+		Certificates: certService,
+		HostResolver: hostResolver,
+	}, nil
+
+}
+
+func (h *Handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log.Info("WEBHOOK HANDLER")
+
+	ingress := &networkingv1.Ingress{}
+	if err := h.Decoder.Decode(req, ingress); err != nil {
+		return admission.Errored(-1, err)
+	}
+
+	original := ingress.DeepCopy()
+
+	allowed, err := h.handle(ctx, ingress)
+	if err != nil {
+		admission.Errored(-1, err)
+	}
+
+	if !allowed {
+		return admission.Denied("")
+	}
+
+	if !equality.Semantic.DeepEqual(original, ingress) {
+		var originalSerialised bytes.Buffer
+		var currentSerialised bytes.Buffer
+
+		h.Serializer.Encode(original, &originalSerialised)
+		h.Serializer.Encode(ingress, &currentSerialised)
+
+		return admission.PatchResponseFromRaw(
+			originalSerialised.Bytes(),
+			currentSerialised.Bytes(),
+		)
+	}
+
+	return admission.Allowed("")
+
+}
+
+func (h *Handler) handle(ctx context.Context, ingress *networkingv1.Ingress) (bool, error) {
+	trafficAccessor := traffic.NewIngress(ingress)
+
+	// verify host is correct
+	// no managed host assigned assign one
+	// create empty DNSRecord with assigned host
+	managedHostRecord, err := h.Hosts.EnsureManagedHost(ctx, trafficAccessor)
```

Comment:
there is a slight change to logic in my PR now in that it returns an array. As thre could be more than one managed host assigned. (most of the time prob only one, but might as well allow for their being more than one)